#456425 [C#テキスト]3章演習問題

kurosu-being opened the pull request:
第3章 ラムダ式とLINQの基礎

1.ラムダ式とはなにか
2.List<T>とラムダ式の組み合わせ
3.LINQ to Objectsの基礎、クエリ演算子、シーケンス
4.即時実行、遅延実行
などを学習しました。

追記(10/24)
第2章演習問題のブランチ(feature/chapter2)から第3章演習問題のブランチを派生させてしまった関係で
当プルリクエストの差分に第2章演習問題の分が出ています。
確認しづらく大変恐縮ですが、ご確認よろしくお願いします。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

第2章演習問題の新規プロジェクトを作成しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

演習問題２の問１-1を書きこみ、1-2の途中までを書き込みました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-1,問1-2を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-3までを実装し、問1-4のファイルを作成しました、

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-4を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1-4においてコーディング規約に則りメソッドにコメントを追加しました

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問2を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問３の途中まで（クラスを書く）を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問3を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問2,問3の問題文を変更、追加しました。
問1のローカル変数song1,song2,song3をコーディング規約に則り変更しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

問1において不要なusingを削除しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

foreach文の中のローカル変数の命名を変更しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

インターフェイスの変数の名前をISale → Saleに変更しました、

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-1-1を実装しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-1-2を実装しました。

kurosu-being committed:
#455475 [C#テキスト]2章演習問題

Practice3-1-3を実装しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-1-4を実装しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

問２のプログラムを作成しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-1を実装しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-2フォルダを作成しました、

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-2を実装しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-3を実装しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-4を実装しました。

yuki-tanaka-being commented on 章末問題3/Practice3-2-1/Program3-2-1.cs:
```diff
@@ -0,0 +1,24 @@
+            var wNames = new List<string> { "Tokyo", "New Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
+            string wInputuser = Console.ReadLine();
+            var wIndex = wNames.FindIndex(s => s == wInputuser);
+            if (wIndex != -1) {
+                Console.WriteLine($"インデックス番号は {wIndex}");
+            } else {
+                Console.WriteLine("-1");
+            }
```
Console.WriteLine() を if-else の中に記載していますが、分岐の中は文字列を構築するだけにし、出力は分岐の外で扱う方が保守性が高くなります。
理由は、出力する文字列を構築する処理と、出力処理を独立させたほうが保守性が高くなるからです。
独立させた方が保守性が高くなるのは、例えば、出力方法を変更するとなった場合、今の処理だと2か所修正が必要です。
また、出力する文字列を構築する条件が増えた場合に、Console.WriteLine() をさらに追加で書く必要があり、この状態で出力処理を変更となったら3か所修正が必要になります。
このように、一つの処理で二つのことをやると、保守性が下がる恐れがあります。
```
            string wOutput = wIndex != -1 ? $"インデックス番号は {wIndex}" : "-1";
            Console.WriteLine(wOutput);
```

ただし、この課題の場合は、そもそも -1 で分岐する必要はないかなと思います。

もう1点、もし分岐する場合の条件は `wIndex != -1` ではなく、`wIndex >= 0` が好ましいです。
理由は、基本的にインデックスを扱うときは 0 以上が存在するを示し、０未満がないことを示します。
ないときは -1 を返す仕様ですが、本質的には -2 でも -100 でもよいです。
インデックスでないを示すのは 0 未満が基本的な条件の扱い方になります。

kurosu-being replied:
ご指摘ありがとうございます！
ご指摘の部分を条件演算子を用いて`wIndex >= 0`の条件で分岐するように変更し、Consoleの処理を独立させました。

> このように、一つの処理で二つのことをやると、保守性が下がる恐れがあります。

確かに、1つの処理が入り組んでしまうとメンテナンスの手間もかかりますし修正の際のヒューマンエラーも誘発させますね。1つの処理＝1つのことでコードが書けるよう練習していきたいと思います👩‍🎓

yuki-tanaka-being reviewed (commented):
ちょっと見てみました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-1において、if文分岐を削除し、条件演算子でインデックスが0以上か判断して出力するように変更しました。

kurosu-being commented:
第2章演習問題のブランチ(feature/chapter2)から現在の第3章演習問題のブランチを派生させてしまった関係で、
当プルリクエストに第2章の変更差分が出てしまっています。
見づらくなってしまっておりますが、ご確認よろしくお願いいたします。

yuki-tanaka-being commented on 章末問題3/Practice3-2-1/Program3-2-1.cs:
```diff
@@ -0,0 +1,21 @@
+            string wInputuser = Console.ReadLine();
+            var wIndex = wNames.FindIndex(s => s == wInputuser);
+            string wMessage = (wIndex >= 0) ? $"インデックス番号は {wIndex}" : "-1";
+            Console.WriteLine($"{wMessage}");
```
`wMessage` は `string` の変数なので、`Console.WriteLine` に渡すときに リテラルに変換せず、そのまま変数を渡せます。
`Console.WriteLine(wMessage);`
推測で記載しますが、もし、元のコードから変更した結果こうなっただけで、最初からこのコードを書いていたらちゃんと書けていたとするなら、
- ひとまず書き終わった後に本当にそれがベストかもう一度ちゃんと見直す
- 修正するときは手をいれる局所的な部分だけでなく周辺も考慮して最適になっているか確認するようにする

といったアクションを実施すると、そういうミスが防ぎやすいかなと思います。

kurosu-being replied:
ご指摘ありがとうございます！直接変数を渡す形に修正いたしました。
ご推察の通り、コピー&ペーストで局部的な手入れを行っていました。アドバイス頂いた2点を、今後特に注意したいと思います。

yuki-tanaka-being reviewed (commented):
修正ありがとうございます。
修正箇所で気になった点をレビューしました。

k-miyahara-being commented on 章末問題3/Practice3-1-1/Program3-1-1.cs:
```diff
@@ -0,0 +1,15 @@
+            //ビーイングのコーディング規約に則り、問題文のnumbersをwNumbersに変更
+            var wNumbers = new List<int> { 12, 87, 94, 14, 53, 20, 40, 35, 76, 91, 31, 17, 48 };
+            var wExsits = wNumbers.Exists(s => s % 9 == 0 || s % 8 == 0);
+            Console.WriteLine(wExsits);
```
誤字です。`wExsits → wExists`
ただ、なくてよい変数ですね！

kurosu-being replied:
ご指摘ありがとうございます！変数名を修正し、不要に定義しているwExsitsを削除しました。

k-miyahara-being commented on 章末問題3/Practice3-1-2/Program3-1-2.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Practice3_1_2 {
+    class Program {
+        //List<T>のメソッドを使い、書く要素を2.0で割った値をコンソールに出力してください。
```
細かいですが、問題文の誤字です。
書く要素 → 各要素

kurosu-being replied:
ご質問ありがとうございます！正しい変換に修正いたしました。

k-miyahara-being commented on 章末問題3/Practice3-1-3/Program3-1-3.cs:
```diff
@@ -0,0 +1,16 @@
+            var wNumbers = new List<int> { 12, 87, 94, 14, 53, 20, 40, 35, 76, 91, 31, 17, 48 };
+            var wWhere = wNumbers.Where(s => s >= 50);
+            foreach (int wN in wWhere) {
+                Console.WriteLine(wN);
```
`wWhere`は不要な変数ですね。
foreach文に直接書いてしまいましょう。
`= wNumbers.Where(s => s >= 50);`の部分が異常に長く読みづらい場合は一旦変数に置いてしまうのもアリと思いますが、今回のような場合はそのまま書いてしまったほうが無難です。

`wN`について、命名し直してみましょう。
第三者がみて意味を理解しづらい変数名になってしまっています。

kurosu-being replied:
ご質問ありがとうございます！
wWhereを削除し、直接書く形に変更いたしました。
また、wNについてもwResultに変更いたしました。

k-miyahara-being commented on 章末問題3/Practice3-1-4/Program3-1-4.cs:
```diff
@@ -0,0 +1,16 @@
+            var wWhere = wNumbers.Select(s => s * 2).ToList();
+            foreach (int wN in wWhere) {
+                Console.WriteLine(wN);
+            }
```
ここは問題に指定があるので、`wWhere`に格納するのは正しいです。
ですが、`wWhere`という変数名も意味が伝わらないものになってしまっているので命名し直してみましょう。

kurosu-being replied:
ご指摘ありがとうございます！
wWhere→wResultに命名を変更いたしました。

k-miyahara-being replied:
`Result`とは「何の」結果か？がわかりづらくなってしまいます。
今回は、その変数を直後の行で使用しているので大きく問題はおこりませんが、定義された箇所から離れた場所でも使用するとすると、そのころには「何の」結果だったっけ？とコードをさかのぼる必要がでてくる可能性があります。

コードのどこからみてもわかるような命名がされていると良いでしょう。
2倍にした結果なので、`wTwiceNumbers`とかですかね。

また、`wMessage`は変数が指すものと異なるように思えます。

kurosu-being replied:
ご質問ありがとうございます。
wResult→wTwiceNumbers
wMessage→wNumber
に変更いたしました。
コードのどこからみてもわかるような命名という観点がなかったので今後意識して命名したいと思います。

k-miyahara-being commented on 章末問題3/Practice3-2-1/Program3-2-1.cs:
```diff
@@ -0,0 +1,21 @@
+        //コーディング規約に則りnames →　wNamesに変更
+        static void Main(string[] args) {
+            var wNames = new List<string> { "Tokyo", "New Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
+            string wInputuser = Console.ReadLine();
```
`wNames`は問題指定の変数名ですが、練習として、ふさわしいものに命名し直してみましょう。
`wInputuser`は変更するとしたら`wUserInput`または単に`wInput`が良いと思いますが、なくてもよいです。

kurosu-being replied:
ご指摘ありがとうございます！
Names→wCityNames
wInputuser→wInput
に変更いたしました。

k-miyahara-being commented on 章末問題3/Practice3-2-1/Program3-2-1.cs:
```diff
@@ -0,0 +1,21 @@
+        //"Tokyo","New Delhi", "Bangkok", "London", "Paris", "Berlin","Canberra"."Hong Kong",
+        //};
+        //このリストに対して、ラムダ式を使い次のコードを書いてください。
+        //Practice3-2-1 コンソールから入力した歳名が何番目に格納されているかList<T>のFindINdexメソッドを使って調べ、その結果をコンソールに出力してください。見つからなかったら、-1を出力してください。なお、コンソールの入力にはConsole.ReadLineメソッドを利用してください。
```
問題文の誤字です。
歳名 → 都市名

プログラムに限ったことではありませんが、誤字は特に気を付けるようにしましょう。
ここはコメントなので影響有りませんが、例えば、画面に表示するメッセージに誤字があると、それがそのままバグとして扱われてしまいます。
もちろんテスト等で他者の目が入るのでそのままお客様のもとに届く確率は低いですが、個人の力で未然に防げるような箇所はなるべく気を付けるようにしておきましょう👍

kurosu-being replied:
コードの中の確認のみで、コメントの確認を怠っていました。
今後商品に携わる為にも誤字がエラーになるということに注意して文章を書くようにします。
ご指摘ありがとうございます！

k-miyahara-being commented on 章末問題3/Practice3-2-2/Program3-2-2.cs:
```diff
@@ -0,0 +1,14 @@
+    class Program {
+        //LINQのCountメソッドを使い、小文字の'o'が含まれている都市名がいくつあるかカウントし、その結果をコンソールに出力してください。
+        static void Main(string[] args) {
+            var wNames = new List<string> { "Tokyo", "ONew Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
```
> "ONew Delhi"

大文字のケースも問題ないかしっかりテスト出来てますね！Goodです。

k-miyahara-being commented on 章末問題3/Practice3-2-3/Program3-2-3.cs:
```diff
@@ -0,0 +1,16 @@
+        //LINQのWhereメソッドを使い、小文字の'o'が含まれている都市名を抽出し、配列に格納してください。その後、配列の各要素をコンソールに出力してください。
+        static void Main(string[] args) {
+            var wNames = new List<string> { "Tokyo", "ONew Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
+            var wWhere = wNames.Where(s => s.Contains("o")).ToArray();
+            foreach (string wN in wWhere) {
+                Console.WriteLine(wN);
```
と同様です。

kurosu-being replied:
ご質問ありがとうございます！
同様に修正を行いました。

k-miyahara-being replied:
リンクはるつもりが忘れてました(笑)
意図くみとっていただきありがとうございます

wMessageという命名と中身が一致していないと思います。

kurosu-being replied:
ご質問ありがとうございます！
Listの名前を、wCityName→wCityNamesに変更し
wMessage→wCityNameに変更しました。
また、csファイル全体の命名を見直し変更しました。

k-miyahara-being commented on 章末問題3/Practice3-2-4/Program3-2-4.cs:
```diff
@@ -0,0 +1,18 @@
+        //LINQのWhereメソッドとSelectメソッドを使い、'B'で始まる都市名の文字数抽出し、その文字数をコンソールに出力してください。都市名を表示する必要はありません。
+        static void Main(string[] args) {
+            var wNames = new List<string> { "Tokyo", "ONew Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
+            var wWhere = wNames.Where(s => s[0] == 'B');
+            var wNum = wWhere.Select(s => s.Length);
```
文字列の先頭の文字を取得するのに、LINQをつかうこともできます。
書き換えてみてください。

また、メソッドチェーンを使って1行で書くことができます。
メソッドチェーンについては、テキストにある(?)と思うので調べてみてください。
なければネットで！

kurosu-being replied:
ご指摘ありがとうございます！
LINQのFirstOrDefaultメソッドを用いる形に変更し、メソッドチェーンを使って書き直しました。

k-miyahara-being replied:
First()ではなくFirstOrDefault()を使用した理由を説明できますか？

kurosu-being replied:
Firstメソッドでは、シーケンスに条件に一致する要素がなかった時例外をスローしますが、
FirstOrDefaultメソッドでは規定値を返します。
条件を検索する場面では、例外を発生させない為にFirstOrDefaultメソッドを使うのが適当だと考えました。

k-miyahara-being replied:
ばっちりですね！
すでに森田さんの指摘があってのFirstOrDefault()だったんですね。
空文字`""`のときに例外が起こることも理解されているようでよかったです！

k-miyahara-being commented on 章末問題3/Practice3-2-4/Program3-2-4.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Practice3_2_4 {
+    class Program {
+        //LINQのWhereメソッドとSelectメソッドを使い、'B'で始まる都市名の文字数抽出し、その文字数をコンソールに出力してください。都市名を表示する必要はありません。
+        static void Main(string[] args) {
```
[追加問題]
同様に、'n' で終わる都市名を抽出し、アルファベット順にならべたものを出力してください。
※ファイルはわけなくてOKです。

kurosu-being replied:
ありがとうございます！追加問題を実装しました。

k-miyahara-being replied:
Orderbyで、先頭文字で並べ替えしていることで、2文字目以降では比較がされず並べ替えされません。
試しに、リストの先頭に"Lyon"を追加して実行してみてください。
↓のようになり、LyonとLondonはアルファベット順としては逆になってしまっています。
```
Edmonton
Lyon
London
```

kurosu-being replied:
ご質問ありがとうございます。OrderBy(x => x)に変更し、全体でアルファベット順になるよう変更いたしました。

s-morita-being commented on 章末問題3/Practice3-2-4/Program3-2-4.cs:
```diff
@@ -0,0 +1,18 @@
+        //LINQのWhereメソッドとSelectメソッドを使い、'B'で始まる都市名の文字数抽出し、その文字数をコンソールに出力してください。都市名を表示する必要はありません。
+        static void Main(string[] args) {
+            var wNames = new List<string> { "Tokyo", "ONew Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
+            var wWhere = wNames.Where(s => s[0] == 'B');
```
この例題では問題にならないので、これでよいと思いますが、
`wNames` に空文字が含まれてもよい仕様の場合、 `s` が空文字だと `s[0]` でエラーが発生します。
ご参考まで。

kurosu-being replied:
ご指摘ありがとうございます！
null条件演算子を用いて空文字の時nullを返すように変更いたしました。

s-morita-being replied:
あ、ちょっと分かりにくい説明でした。
空文字というのは `null` ではなくて `""` のことで、↓の場合に `s[0]` でインデックスエラーが発生する。という意味でした。
```
var wNames = new List<string> { "" };
```
ので、既に修正してもらっている `s.FirstOrDefault() == 'B'` で解消しているはずです。
(`s.Length > 0 && s[0] == 'B'` でも OK です。)

逆に wNames が null になることはないので、 `?.` の `?` はない方がよいです。

文字列は `null` と空文字 `""` で違うものなので注意してください。
仕様や設計によって、両者に異なる意味が与えられていて明確に区別しないといけない時もあれば、
区別があいまいで同一扱いした方がよい時もあったりします。

kurosu-being replied:
ご指摘の意味を誤解しておりました、申し訳ないです🙇‍♀️
?.を削除いたしました。

先頭字（s[0]）が存在しない文字列が存在しうる　→　それが空文字 `""`ということですね。
` null` と空文字 `""`について混同して理解していた部分があったので、改めて調べようと思います。
ご指摘ありがとうございます！

s-morita-being reviewed (commented):
お疲れ様です。
ひとつコメントしました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

全てのcsファイルにおいて、不要な変数の削除、不十分な命名（wNなど）を変更しました。
Practice3-2-4において、LINQとメソッドチェーンを用いて書き直しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-4において追加問題を修正しました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-1-4を問題文の指定通りに変更いたしました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

?.を削除いたしました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

ファイル全体で変数の命名を変更しました。
Practice3-2-4において、OrderByのラムダ式を変更しました、

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

不要なコメントを削除しました。

k-miyahara-being commented on 章末問題3/Practice3-2-1/Program3-2-1.cs:
```diff
@@ -0,0 +1,21 @@
+            var wCityNames = new List<string> { "Tokyo", "New Delhi", "Bangkok", "London", "Paris", "Berlin", "Canberra", "Hong Kong" };
+            string wInput = Console.ReadLine();
+            var wIndex = wCityNames.FindIndex(s => s == wInput);
+            string wMessage = (wIndex >= 0) ? $"インデックス番号は {wIndex}" : "-1";
+            Console.WriteLine(wMessage);
```
リストに存在しないとき、「-1」と表示されてもユーザーは何がなんだか分からないと思います。
対話型の場合は、ユーザーが迷わないようにすることが大切です。
メッセージを考えて書き直してみましょう！

kurosu-being replied:
承知しました。
「その都市は登録されていません」にメッセージを変更いたしました。

kurosu-being committed:
#456425 [C#テキスト]3章演習問題

Practice3-2-1においてメッセージを変更しました。

k-miyahara-being reviewed (approved):
OKです！
